controllers/perconaservermysqlbackup_controller.go

```diff
@@ -315,3 +328,200 @@ func (r *PerconaServerMySQLBackupReconciler) getBackupSource(ctx context.Context
+		if err := r.Get(ctx, nn, s); err != nil {
+			return nil, errors.Wrapf(err, "get secret/%s", nn.Name)
+		}
+		storageAccount, ok := s.Data[secret.CredentialsAzureStorageAccount]
+		if !ok {
+			return nil, errors.Errorf("no credentials for Azure in secret %s", nn.Name)
+		}
+		accessKey, ok := s.Data[secret.CredentialsAzureAccessKey]
+		if !ok {
+			return nil, errors.Errorf("no credentials for Azure in secret %s", nn.Name)
+		}
+		conf.Azure.ContainerName = azure.ContainerName
+		conf.Azure.EndpointURL = azure.EndpointURL
+		conf.Azure.StorageClass = azure.StorageClass
+		conf.Azure.StorageAccount = string(storageAccount)
+		conf.Azure.AccessKey = string(accessKey)
+		conf.Type = apiv1alpha1.BackupStorageAzure
+	default:
+		return nil, errors.New("unknown backup storage type")
+	}
+	return conf, nil
+}
+
+func (r *PerconaServerMySQLBackupReconciler) deleteBackup(ctx context.Context, cr *apiv1alpha1.PerconaServerMySQLBackup) (bool, error) {
+	backupConf, err := r.backupConfig(ctx, cr)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to create sidecar backup config")
+	}
+
+	cluster := new(apiv1alpha1.PerconaServerMySQL)
+	nn := types.NamespacedName{Name: cr.Spec.ClusterName, Namespace: cr.Namespace}
+	err = r.Client.Get(ctx, nn, cluster)
+	if client.IgnoreNotFound(err) != nil {
+		return false, errors.Wrapf(err, "get cluster %s", nn)
+	}
+	if k8serrors.IsNotFound(err) {
+		job := &batchv1.Job{}
+		nn = types.NamespacedName{Name: xtrabackup.DeleteJobName(cr), Namespace: cr.Namespace}
+		err = r.Client.Get(ctx, nn, job)
+		if client.IgnoreNotFound(err) != nil {
+			return false, errors.Wrapf(err, "get job %s", nn)
+		}
+		if k8serrors.IsNotFound(err) {
+			job = xtrabackup.GetDeleteJob(cr, backupConf)
+			if err := controllerutil.SetControllerReference(cr, job, r.Scheme); err != nil {
+				return false, errors.Wrapf(err, "set controller reference to Job %s/%s", job.Namespace, job.Name)
+			}
+
+			if err := r.Client.Create(ctx, job); err != nil {
+				return false, errors.Wrapf(err, "create job %s/%s", job.Namespace, job.Name)
+			}
+			return false, nil
+		}
+
+		complete := false
+		for _, cond := range job.Status.Conditions {
+			if cond.Status != corev1.ConditionTrue {
+				continue
+			}
+
+			switch cond.Type {
+			case batchv1.JobFailed:
+				return false, errors.New("job failed")
+			case batchv1.JobComplete:
+				complete = true
+			}
+		}
+		return complete, nil
+	}
+	data, err := json.Marshal(backupConf)
+	if err != nil {
+		return false, errors.Wrap(err, "marshal sidecar backup config")
+	}
+	src, err := r.getBackupSource(ctx, cluster)
+	if err != nil {
+		return false, errors.Wrap(err, "get backup source node")
+	}
+	sidecarURL := url.URL{
+		Host:   src + ":6033",
+		Scheme: "http",
+		Path:   "/backup/" + cr.Name,
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, sidecarURL.String(), bytes.NewReader(data))
+	if err != nil {
+		return false, errors.Wrap(err, "create http request")
+	}
```

Comment:
I don't understand why we create a job to delete backup and also send DELETE request to sidecar. They seem to run the same command. What am I missing here?

Reply:
@egegunes ff a cluster is deleted, there are no available pods to run `xbcloud`, that's why we create a job. Otherwise, we don't need an additional pod and we send `DELETE` to the available pod

Reply:
I see, thanks for explanation